### PR TITLE
Adding adapter support for NeoX

### PR DESCRIPTION
**Pull request by ajesujoba**
Added adapter support for GPTNeoX  with tests. Although at the moment, while training the adapter module, it also trains the CLM head (which is not the expected situation). This I already raised [here](https://github.com/adapter-hub/adapter-transformers/issues/521).

### CODE REVIEW

File: src/transformers/__init__.py [hunk 1]
```diff
@@ -2942,6 +2944,15 @@
...
     )
+    _import_structure["models.gpt_neox"].extend(
+        [
+            "TFGPTNeoXForCausalLM",
+            "TFGPTNeoXForQuestionAnswering",
+            "TFGPTNeoXForSequenceClassification",
+            "TFGPTNeoXModel",
+            "TFGPTNeoXPreTrainedModel",
+        ]
+    )
```
Comment (calpt) [inline src/transformers/__init__.py:2955]: Is the addition of these imports related to changes of this PR?

File: src/transformers/adapters/models/gpt_neox/adapter_model.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        num_labels=2,
+        layers=1,
+        activation_function="tanh",
+        overwrite_ok=False,
+        id2label=None,
+    ):
+        head = QuestionAnsweringHead(self, head_name, num_labels, layers, activation_function, id2label)
+        self.add_prediction_head(head, overwrite_ok)
+
+class GPTNeoXModelWithHeads(GPTNeoXAdapterModel):
```
Comment (calpt) [inline src/transformers/adapters/models/gpt_neox/adapter_model.py:188]: As XModelWithHeads classes are deprecated, we wouldn't want to add this class for newly supported models.

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 3]
```diff
@@ -95,7 +105,7 @@ def __init__(self, config):
             self.rotary_ndims, config.max_position_embeddings, base=config.rotary_emb_base
         )
         self.norm_factor = torch.sqrt(torch.tensor(self.head_size, dtype=torch.float32)).to(torch.get_default_dtype())
-        self.query_key_value = nn.Linear(config.hidden_size, 3 * config.hidden_size)
+        self.query_key_value = LoRALinear(config.hidden_size, 3 * config.hidden_size, "selfattn", config)
```
Comment (calpt) [inline src/transformers/models/gpt_neox/modeling_gpt_neox.py:108]: As GPT-NeoX groups all attention matrices together into one linear layers similar to GPT-2, this should probably use `LoRAMergedLinear` instead of `LoRALinear` if i'm not mistaken (see GPT-2 implementation for reference).

**Review by calpt — COMMENTED** [review]
@ajesujoba Sorry for the delay in getting back to you. I've done an initial review of your changes and left some comments. Overall, this looks very promising, thanks again for your efforts.

Before this can be merged, please additionally make sure:
- all checks and tests in the CI pipeline pass
- all points in [our contributing guide](https://docs.adapterhub.ml/contributing/adding_adapters_to_a_model.html) are addressed
- in `tests_adapters/models`, a new module for GPT-NeoX is added (this is currently not correctly documented in the contributing guide, we'll update)

Additionally, I've provided a fix for the issue regarding output embeddings you've raised and we've discussed in #537.

File: src/transformers/__init__.py (hunk 1 above)
Comment (ajesujoba) [reply src/transformers/__init__.py:2955]: No, they are not. Should I remove them

File: src/transformers/adapters/models/gpt_neox/adapter_model.py (hunk 2 above)
Comment (ajesujoba) [reply src/transformers/adapters/models/gpt_neox/adapter_model.py:188]: Sure, we can remove them.

Comment (ajesujoba) [reply src/transformers/adapters/models/gpt_neox/adapter_model.py:188]: @calpt , one more thing to point out. [GPTNeoX configuration](https://github.com/adapter-hub/adapter-transformers/blob/master/src/transformers/models/gpt_neox/configuration_gpt_neox.py) does not include `'hidden_dropout_prob' ` and `attention_probs_dropout_prob`. And running test_adapters with this would result in an error. What do you suggest?

File: src/transformers/__init__.py (hunk 1 above)
Comment (ajesujoba) [reply src/transformers/__init__.py]: No, I removed them

Comment (ajesujoba) [reply src/transformers/__init__.py]: Regarding `all points in [our contributing guide](https://docs.adapterhub.ml/contributing/adding_adapters_to_a_model.html) are addressed`, it is optional in the documentation and I tried it but having some mismatches in dimension, Is it still optional?

Comment (calpt) [reply src/transformers/__init__.py]: Could you clarify what exactly you're referring to from the contributing guide? The Parallel inference and static head conversion points are still optional (although highly recommended). If Parallel support is not implemented, please make sure to remove the test mixin classes starting with "Parallel..." from the model test class.

File: src/transformers/adapters/models/gpt_neox/adapter_model.py (hunk 2 above)
Comment (calpt) [reply src/transformers/adapters/models/gpt_neox/adapter_model.py]: Regarding the tokenizer, I think we should we should allow using fast tokenizers anyway. Could you check if/ which issues occur when you switch to loading the fast tokenizer for GPTNeoX?

Regarding the dropout issue, we would probably add some default dropout if the keys are not present in the model config. I can look into this. You could just add the missing key to the model config temporarily to make the tests pass and we can change it later on.

File: src/transformers/__init__.py (hunk 1 above)
Comment (ajesujoba) [reply src/transformers/__init__.py]: Yes, I was referring to the Parallel inference and static head conversion points are still optional (although highly recommended). And as you recommended I would remove test mixin classes starting with "Parallel...". I would also remove some of the other tests such as `IA3TestMixin, LoRATestMixin, PrefixTuningTestMixin, UniPELTTestMixin,` as they require adding classification head `'add_classification_head'`  and the GPT_NeoX model in this version does not have that .

File: src/transformers/adapters/models/gpt_neox/adapter_model.py (hunk 2 above)
Comment (ajesujoba) [reply src/transformers/adapters/models/gpt_neox/adapter_model.py]: Ok! Sounds good

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py (hunk 3 above)
Comment (ajesujoba) [reply src/transformers/models/gpt_neox/modeling_gpt_neox.py]: I fixed this already!

**Comment by calpt** [comment]
Hey, thanks again for you efforts in contributing new model architectures to adapter-transformers and sorry for the silence on our side.

In the last few weeks, we've been working on a large refactoring of our project, which will ultimately result in the release of _Adapters_, the next-generation adapters library. See https://github.com/adapter-hub/adapter-transformers/issues/584.

As a consequence, we plan to merge any new model integrations directly to the new codebase, which currently can be found [on this branch](https://github.com/adapter-hub/adapter-transformers/tree/adapters). Unfortunately, this necessitates some changes in the model integration code (detailed [here](https://github.com/adapter-hub/adapter-transformers/blob/adapters/docs/contributing/adding_adapters_to_a_model.md), see already integrated models such as BERT, BART etc. for reference).

If you'd be willing to update your model integration to target the new library yourself, we'd be super happy to help you on this. Otherwise, we might look into upgrading and merging some of the open model integration PRs ourselves in the future. For more details, again see https://github.com/adapter-hub/adapter-transformers/issues/584.
